### ISSUE TITLE
Compilation errors when defining mutually recursive models

### ISSUE THREAD

**Issue by numso** [zookzook/elixir-mongodb-driver#199]
## The issue

I am getting compilation errors when trying to create models for mutually recursive data.

Is there a way to create these models today without errors? If not, is this a feature you'd be willing to support in your library? Or is this out of scope for this project?

## Problem details

Some of the data I have in mongo is stored something like this:

```elixir
%{
  _id: "some_id",
  steps: [
    %{
      details: %{}
    },
    %{
      details: %{},
      subflows: [
        %{
          rules: %{},
          steps: [...]
        }
      ]
    }
  ]
}
```

It's an object that has A's which can contain B's which can contain A's.
In my case I have Steps which can contain Subflows which can contain Steps.

## What I've tried

I've tried modeling it like so:
```elixir
# workflow.ex
defmodule Workflow do

  defmodule Subflow do
    use Mongo.Collection
    document do
      attribute :rules, map()
      embeds_many :steps, Step
    end
  end

  defmodule Step do
    use Mongo.Collection
    document do
      attribute :details, map()
      embeds_many :subflows, Subflow
    end
  end

  use Mongo.Collection
  collection "workflows" do
    embeds_many :steps, Step
  end
end
```

but that results in compilation errors like these:
```
== Compilation error in file lib/workflow.ex ==
** (UndefinedFunctionError) function Step.__info__/1 is undefined (module Step is not available)
    Step.__info__(:functions)
    (mongodb_driver 1.0.3) lib/mongo/collection.ex:844: Mongo.Collection.has_load_function?/1
    (elixir 1.14.4) lib/enum.ex:4197: Enum.filter_list/2
    (elixir 1.14.4) lib/enum.ex:4198: Enum.filter_list/2
    lib/workflow.ex:5: (module)
```

If I break those modules out into different files like this:
```elixir
# workflow.ex
defmodule Workflow do
  use Mongo.Collection
  collection "workflows" do
    embeds_many :steps, Step
  end
end

# subflow.ex
defmodule Subflow do
  use Mongo.Collection
  document do
    attribute :rules, map()
    embeds_many :steps, Step
  end
end

# step.ex
defmodule Step do
  use Mongo.Collection
  document do
    attribute :details, map()
    embeds_many :subflows, Subflow
  end
end
```

then I get this error instead:
```
== Compilation error in file lib/step.ex ==
** (CompileError)  deadlocked waiting on module Subflow
    (mongodb_driver 1.0.3) lib/mongo/collection.ex:844: Mongo.Collection.has_load_function?/1
    (elixir 1.14.4) lib/enum.ex:4197: Enum.filter_list/2
    (elixir 1.14.4) lib/enum.ex:4198: Enum.filter_list/2
    lib/step.ex:3: (module)

== Compilation error in file lib/subflow.ex ==
** (CompileError)  deadlocked waiting on module Step
    (mongodb_driver 1.0.3) lib/mongo/collection.ex:844: Mongo.Collection.has_load_function?/1
    (elixir 1.14.4) lib/enum.ex:4197: Enum.filter_list/2
    (elixir 1.14.4) lib/enum.ex:4198: Enum.filter_list/2
    lib/subflow.ex:3: (module)

Compilation failed because of a deadlock between files.
The following files depended on the following modules:

  lib/step.ex => Subflow
        lib/subflow.ex => Step

Ensure there are no compile-time dependencies between those files and that the modules they reference exist and are correctly named
```

**Comment by zookzook** [comment]
> Ensure there are no compile-time dependencies between those files and that the modules they reference exist and are correctly named

But the modules have dependencies between themself and that's the reason for the deadlock. You need to "break" the cyclic reference: Instead of using the `embeds_many :subflows, Subflow` macro you specify a simple attribute `attribute :subflows, list(map()), default []` and the `dump/1` and `load/1` functions implement the serialization of the `subflows`.

Example:
```elixir
defmodule Step do
  use Mongo.Collection
  document do
    attribute :details, map()
    attribute :subflows, list(map())
    
    after_load &Step.after_load/1
    before_dump &Step.before_dump/1
  end
  
  def after_load(step) do
    Map.put(step, :subflows, Enum.map(step.subflows || [], fn subflow -> Subflow.load(subflow) end)
  end

  def before_dump(step) do
    Map.put(step, :subflows, Enum.map(step.subflows || [], fn subflow -> Subflow.dump(subflow) end)
  end
end
```

**Comment by numso** [comment]
Ah, thank you! Your explanation totally makes sense and that example works perfectly.
